### ISSUE TITLE
Query: AQL: missing variable # for node #... location RestCursorHandler.cpp

### ISSUE THREAD

**Issue by hwde** [arangodb/arangodb#14819]
## My Environment

* __ArangoDB Version__:        3.8.1
* __Storage Engine__:             RocksDB
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:   ArangoDB Starter in Docker
* __Configuration__:               -
* __Infrastructure__:               own
* __Operating System__:        MacOS 11.6
* __Total RAM in your machine__:        
* __Disks in use__:        SSD
* __Used Package__:             image: arangodb:3.8

## Component, Query & Data

__Affected feature__:
__AQL query (if applicable)__:
```
for doc in UNION(
  (
    for doc in @@impressionCollection1
      FILTER doc.dt >= @from AND doc.dt < @to 
      COLLECT fr_fd_date=DATE_FORMAT(DATE_TRUNC(doc.dt,"day"),"%yyyy-%mm-%dd") 
      AGGREGATE fr_fr_views=SUM(doc.im * doc.au) 
      LIMIT 1000000
      RETURN { fr_fd_date,fr_fr_views }
  ),[])
 LIMIT 1000000
return doc
```
__AQL explain and/or profile (if applicable)__:
__Dataset__:
Collection with one document should be enought:

{
  "dt": 1597676916933,
  "au": 17.0111,
  "im": 1
 }

__Size of your Dataset on disk__:
N/A

__Replication Factor & Number of Shards (Cluster only)__:

## Steps to reproduce

1. Run AQL query
2. 
3. 

__Problem__:

Query: AQL: missing variable #17 (16) for node #6 (CalculationNode) while planning registers (exception location: /work/ArangoDB/arangod/Aql/RegisterPlan.cpp:176). Please report this error to arangodb.com (while optimizing plan) (exception location: /work/ArangoDB/arangod/RestHandler/RestCursorHandler.cpp:290). Please report this error to arangodb.com

__Expected result__:

Empty or some results.

**Comment by jsteemann** [comment]
@hwde: could you also supply the exact bind parameter values and which exact indexes are present on the underlying collections? It may have to do with that, because I tried the query locally and didn't run into this problem. Thanks!

**Comment by hwde** [comment]
My JSON bind-values are
```
{
	"@impressionCollection1": "foo",
	"from": 1596240000000,
	"to": 1596412800000
}
```
and the collections structure from a dump is:

```
{"indexes":[{"id":"6027727","type":"persistent","name":"idx_1644469791088443392","fields":["ci"],"unique":false,"sparse":false,"deduplicate":true,"estimates":true},{"id":"6027731","type":"persistent","name":"idx_1644469791089491968","fields":["pi","dt"],"unique":false,"sparse":false,"deduplicate":true,"estimates":true},{"id":"6027735","type":"persistent","name":"idx_1644469791089491969","fields":["dt","pi"],"unique":false,"sparse":false,"deduplicate":true,"estimates":true}],"parameters":{"allowUserKeys":true,"cacheEnabled":false,"cid":"198","deleted":false,"globallyUniqueId":"h1B711C0145F8/6027722","id":"198","isDisjoint":false,"isSmart":false,"isSmartChild":false,"isSystem":false,"keyOptions":{"allowUserKeys":true,"type":"traditional","lastValue":629451},"minReplicationFactor":1,"name":"foo","numberOfShards":1,"planId":"6027722","replicationFactor":1,"schema":null,"shardKeys":["_key"],"shards":{},"status":3,"syncByRevision":false,"type":2,"usesRevisionsAsDocumentIds":false,"version":9,"waitForSync":false,"writeConcern":1}}
```

I get the error even if I truncate the collection or drop the docker volume and restored from a previous dump.

**Comment by hwde** [comment]
Now I was able to reproduce with a newly created collection:
- Open UI, add a collection "foo"
- go to "Indexes" of foo, click "add"
- select "primary", set field to "dt" and name to "dt", click "Create"
- run query.

**Comment by jsteemann** [comment]
Fixed via the 2 above PRs, targeting 3.8.2 and 3.9.0.

**Comment by maxkernbach** [comment]
Hi @hwde,

I am closing this issue since a fix is available in latest versions 3.7.17, 3.8.6 and 3.9.0.
In case an error still occurs, please comment or open a new issue.
